### PR TITLE
Update parity-codec version

### PR DESCRIPTION
**Pull request by DarkEld3r**

### CODE REVIEW

File: primitive-types/impls/codec/Cargo.toml [hunk 1]
```diff
@@ -1,13 +1,13 @@
 [package]
 name = "impl-codec"
```
Comment (dvdplm) [inline primitive-types/impls/codec/Cargo.toml]: Does this need a rename too?

File: primitive-types/impls/codec/Cargo.toml [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
-version = "0.2.0"
+version = "0.3.0"
 authors = ["Parity Technologies <admin@parity.io>"]
 license = "Apache-2.0/MIT"
 homepage = "https://github.com/paritytech/parity-common"
 description = "Parity Codec serialization support for uint and fixed hash."
 
 [dependencies]
-parity-codec = { version = "3.0", default-features = false }
+parity-codec = { version = "4.0", default-features = false }
```
Comment (dvdplm) [inline primitive-types/impls/codec/Cargo.toml]: Did the crate name stay the same?

Comment (DarkEld3r) [reply primitive-types/impls/codec/Cargo.toml]: Yes for now. We currently have 4.0 version of `parity-codec`, it is renamed to `parity-scale-codec` and will be released as `1.0.0` in the future. So we need to break everything twice.

**Comment by DarkEld3r** [comment]
@dvdplm Makes sense. I'm going to create pull requests to `substrate` and `parity-ethereum` referencing this branch.

**Comment by dvdplm** [comment]
Feel free to publish beta versions of the crates as needed: `version = "x.y.z.beta.1`.

**Comment by dvdplm** [comment]
@DarkEld3r I'll put this "in progress" for the time being to avoid accidental merges. :)

**Comment by bkchr** [comment]
As this is just a breaking release, because we removed the implementation for usize. I don't think that we need expensive testing in front of releasing these new versions.

**Comment by thiolliere** [comment]
true if it works with latest 3.5.1, then only change is usize removal.

You can do a substrate branch with patch dependencies to ensure this breaks doesn't create troubles, but as usize usage was wrong, It shouldn't add any trouble actually.

File: ethereum-types/Cargo.toml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "ethereum-types"
-version = "0.7.0"
+version = "0.8.0"
```
Comment (ordian) [inline ethereum-types/Cargo.toml]: note, that we haven't published ethereum-types 7.0, so no need to version bump here

File: keccak-hash/Cargo.toml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "keccak-hash"
-version = "0.3.0"
+version = "0.4.0"
```
Comment (ordian) [inline keccak-hash/Cargo.toml]: same

File: primitive-types/Cargo.toml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "primitive-types"
-version = "0.4.0"
+version = "0.5.0"
```
Comment (ordian) [inline primitive-types/Cargo.toml]: ditto

**Comment by DarkEld3r** [comment]
Here is an almost green (I suppose `impl_version` should be updated?) pull request to substrate: https://github.com/paritytech/substrate/pull/2855

Additional updates were required:
- https://github.com/paritytech/finality-grandpa/pull/68
- https://github.com/paritytech/rhododendron/pull/22
- `trie-bench` and `reference-trie`: https://github.com/paritytech/trie/pull/19

**Pull request merged**